lib/tagname.c

```diff
@@ -61,21 +55,31 @@ static int tagCmpValue(const void * avp, const void * bvp)
     return ret;
 }
 
-static pthread_once_t tagsLoaded = PTHREAD_ONCE_INIT;
+struct tagTable {
+    private:
+    headerTagTableEntry byName[TABLESIZE]; /*!< tags sorted by name */
+    headerTagTableEntry byValue[TABLESIZE]; /*!< tags sorted by value */
 
-/* Initialize tag by-value and by-name lookup tables */
-static void loadTags(void)
+    public:
+    tagTable();
+    headerTagTableEntry getEntry(const char *tag);
+    headerTagTableEntry getEntry(rpmTagVal tag);
+    int getNames(rpmtd tagnames, int fullname);
+};
```

Comment:
How does this work? I didn't think `struct` supported `private`/`protected`/`public`?

Reply:
It does, the difference to class is just the the default (and inheritance): struct defaults to everything public, class defaults to everything private. Using struct here is silly here of course, since it has private members it might as well be a class. It's a struct because ... in the first work-in-progress version those parts needed to be public and out of sheer habit I guess :sweat_smile: Not that it matters much either way because it's just this hidden thing inside a single source.

Another potential reason (generally, not here) for preferring struct over class might be C-compatibility: C will take "typedef struct foo *bar" but not class. I haven't tried but I doubt C++ will let you interchange class/struct as a pointer type.

Reply:
That actually raises another largely stylistic point: for whatever reason (could be Turbo Pascal for all I know), I tend to put private members first and public last. Which is bonkers - the user of your class will primarily/only be interested in the *public* parts, so they should come first, really. OTOH my habit is equally bonkers as the class encapsulation of C++ in general: you need to place the guts of your classes on display and then just put a glass in between: 'na-ah, you can look but you can't touch'. Or add a second wrapper object with all the human and computer overhead just to hide that - the information hiding touted by OO languages, C does it better for public APIs.

Reply:
> Another potential reason (generally, not here) for preferring struct over class might be C-compatibility: C will take "typedef struct foo *bar" but not class. I haven't tried but I doubt C++ will let you interchange class/struct as a pointer type.

Actually, it does. It seems that the only place where struct/class makes a difference to C++ is the actual definition. Otherwise at least gcc seems to let to you interchange struct/class with no effect or complaint :flushed: 
Handy for C-compatibility but perhaps confusing from the codebase POV: you grep for "struct foo", can't find it because it's defined as "class foo", and then scratch head because somehow it works nevertheless :sweat_smile:

Reply:
Can we just use `class` then? Because it's a little weird to see `struct` used this way.

Reply:
I'm curious, what exactly is so weird about it, to you?

I totally agree this thing here is a class, and we can call it that way because in this particular case it makes no difference either way. But looking at the bigger picture, I find it far, far more confusing to have the same thing called a struct in one place and class in another so grep wont find it trivially. And we're going to have a lot of that.

Reply:
I guess for me, I see `struct` as a way to do C-style data-only layouts. As soon as you add methods to the thing, I would expect it to be a `class`.

I am aware that classes are basically fancy structs with extra sugar, but the way I (and I think most people) expect to look at a C++ codebase is that `struct` is data-only while `class` can have data + methods.

Reply:
This silly struct/class issue is of course really way beyond the scope of this particular PR.

Common style aside, in the RPM context it comes down to: can we accept a situation where half of our codebase talks about "struct rpmfoo" and the other half talks about "class rpmfoo", while both mean the same exact thing?
If the answer is no, and all of our major items like rpmts, rpmfi, rpmds etc are condemned to be structs despite having methods, then calling some individual item here class will make that *class* look odd :sweat_smile:

Reply:
Also, with rolling over to a new major, we can change the C API that we expose somewhat too.

Reply:
rpmts and all that stuff *will* grow methods sooner or later. If we follow the strict methods == class norm, we *will* also have the madness where it's called struct in one place and class in another. How big an issue that is in practise, I don't know. It seems pretty hideous on the outset, but its quite a mess either way.

Except for the odd addition here or there, changing C API is way out of scope for 6.0, there's neither time, will or energy to do anything about that just now.

Reply:
I'm of the opinion (largely influenced by the little I've read on this topic so far) that for entities whose primary purpose is to be an *aggregate* of data, we should just use structs. Even if we enrich them with private members and public getters/setters.

It does go against the common notion of a class in other, strictly OOP, languages, but we could also look at this as an advantage of the flexibility of C++ :smile:. In those other languages, you often end up with these awkward "classes" that really only hold data and have no internal logic, nor are they used for inheritance. Yes, some languages like Python have tuples or dictionaries (which are still objects but not *classes*) but some only have classes.

`tagTable` is one such example; it *is* a bundle of data, with some very little logic sprinkled on top that initializes it and provides access to its data in different ways.

That said, if we were to follow the "invariant -> class" rule, then `tagTable` might as well be a class :sweat_smile:. That's because there is relation between its members (`byName[]` and `byValue[]`) in the sense that they both use the same underlying data (`rpmTagTable[]`). But I wouldn't sweat this.

Reply:
> Common style aside, in the RPM context it comes down to: can we accept a situation where half of our codebase talks about "struct rpmfoo" and the other half talks about "class rpmfoo", while both mean the same exact thing? 

What do you mean exactly? Why would we refer to `rpmfoo` as a struct in one place and a class in another? Any example in mind? :sweat_smile:

Reply:
Thinking about it a bit more, using both structs and classes for what most programmers consider *classes* takes its toll... So I'd say, let's not try to decide *here* but keep that in mind going forward.

Reply:
> What do you mean exactly? Why would we refer to rpmfoo as a struct in one place and a class in another? Any example in mind?

Because a C compiler does not understand the "class" keyword, yet we need to be able to pass pointers to our classes over to C. Consider the ubiquitous "rpmts" - it's defined as "typedef struct rpmts_s * rpmts;", similar opaque pointer handles exist for all our main data types. The underlying rpmts_s will sooner or later grow methods, and if we make it into a class at that point, as per the above, we'll have that class/struct disparency in large parts of the codebase. The compiler doesn't mind, but humans might :sweat_smile:

Reply:
Changed to class for now, we can ponder more the meaning of life and everything in some other context :smile:

Reply:
> Because a C compiler does not understand the "class" keyword, yet we need to be able to pass pointers to our classes over to C.
[...]

Oh, that explains it, thanks!